modules/drivers/snowflake/src/metabase/driver/snowflake.clj

```diff
@@ -227,69 +227,68 @@
-                              (hx//
-                               (hx/-
-                                (hsql/call :datediff (hsql/raw "month") a b)
-                                (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b))))
-                               3))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :month
-      (let [positive-diff (fn [a b]
-                            (hx/-
-                             (hsql/call :datediff (hsql/raw "month") a b)
-                             (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b)))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :week
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (hsql/call :datediff (hsql/raw "day") a b) 7))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x))))
-
-      :day
-      (hsql/call :datediff (hsql/raw (name unit)) x y)
-
-      (:hour :minute :second)
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (hx/cast :float (hsql/call :datediff (hsql/raw "millisecond") a b)) ; datediff returns integer, so cast to float
-                                    (case unit :hour 3600000 :minute 60000 :second 1000)))))]
-        (hsql/call :case (hsql/call :<= x y) (positive-diff x y) :else (hx/* -1 (positive-diff y x)))))))
+(defn- time-zoned-datediff
+  "Same as snowflake's `datediff` but converts the args to the results time zone
+   before calculating date boundaries. This is needed when an argument could be of
+   timestamptz type and the unit is day, week, month, quarter or year."
+  [unit x y]
+  (let [x (cond->> x
+            (hx/is-of-type? x "timestamptz")
+            (hsql/call :convert_timezone (qp.timezone/results-timezone-id)))
+        y (cond->> y
+            (hx/is-of-type? y "timestamptz")
+            (hsql/call :convert_timezone (qp.timezone/results-timezone-id)))]
+    (hsql/call :datediff (hsql/raw (name unit)) x y)))
+
+(defn- time-zoned-extract
+  "Same as `extract` but converts the arg to the results time zone if it's a timestamptz."
+  [unit x]
+  (let [x (cond->> x
+            (hx/is-of-type? x "timestamptz")
+            (hsql/call :convert_timezone (qp.timezone/results-timezone-id)))]
+    (extract unit x)))
+
+(defn- sub-day-datediff
+  "Same as snowflake's `datediff`, but accurate to the millisecond for sub-day units."
+  [unit x y]
+  (let [milliseconds (hx/cast :float (hsql/call :datediff (hsql/raw "milliseconds") x y))]
+    ; millseconds needs to be cast to float because division rounds incorrectly with large integers
```

Comment:
https://guide.clojure.style/#two-semicolons-for-code-fragment

Reply:
Updated https://github.com/metabase/metabase/commit/e6103d45832615f540483bb6477d3e4214cf996a. Also noticed I was casting to float twice.